pallets/moonbeam-lazy-migrations/src/lib.rs

```diff
@@ -65,6 +86,209 @@ pub mod pallet {
+		}
+
+		pub fn add_reads(&mut self, reads: u64) {
+			self.reads += reads;
+		}
+
+		pub fn add_writes(&mut self, writes: u64) {
+			self.writes += writes;
+		}
+	}
+
+	#[derive(Clone)]
+	struct MigrationResult {
+		last_key: Option<StorageKey>,
+		error: Option<&'static str>,
+		reads: u64,
+		writes: u64,
+	}
+
+	enum NextKeyResult {
+		NextKey(StorageKey),
+		NoMoreKeys,
+		Error(&'static str),
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Handle the migration of the storage keys, returns the number of read and write operations
+		fn handle_migration(remaining_weight: Weight) -> ReadWriteOps {
+			let mut read_write_ops = ReadWriteOps::new();
+
+			// maximum number of items that can be migrated in one block
+			let migration_limit = remaining_weight
+				.proof_size()
+				.saturating_sub(PROOF_SIZE_BUFFER)
+				.saturating_div(MAX_ITEM_PROOF_SIZE);
+
+			let status = StateMigrationStatus::<T>::get();
+			read_write_ops.add_one_read();
+
+			let next_key = match &status {
+				MigrationStatus::NotStarted => Default::default(),
+				MigrationStatus::Started(storage_key) => {
+					let (reads, next_key_result) = Pallet::<T>::get_next_key(storage_key);
+					read_write_ops.add_reads(reads);
+					match next_key_result {
+						NextKeyResult::NextKey(next_key) => next_key,
+						NextKeyResult::NoMoreKeys => {
+							StateMigrationStatus::<T>::put(MigrationStatus::Complete);
+							read_write_ops.add_one_write();
+							return read_write_ops;
+						}
+						NextKeyResult::Error(e) => {
+							StateMigrationStatus::<T>::put(MigrationStatus::Error(
+								e.as_bytes().to_vec().try_into().unwrap_or_default(),
+							));
+							read_write_ops.add_one_write();
+							return read_write_ops;
+						}
+					}
+				}
+				MigrationStatus::Complete | MigrationStatus::Error(_) => {
+					return read_write_ops;
+				}
+			};
+
+			let res = Pallet::<T>::migrate_keys(next_key, migration_limit);
+			read_write_ops.add_reads(res.reads);
+			read_write_ops.add_writes(res.writes);
+
+			match (res.last_key, res.error) {
+				(None, None) => {
+					StateMigrationStatus::<T>::put(MigrationStatus::Complete);
+					read_write_ops.add_one_write();
+				}
+				// maybe we should store the previous key in the storage as well
+				(_, Some(e)) => {
+					StateMigrationStatus::<T>::put(MigrationStatus::Error(
+						e.as_bytes().to_vec().try_into().unwrap_or_default(),
+					));
+					read_write_ops.add_one_write();
+				}
+				(Some(key), None) => {
+					StateMigrationStatus::<T>::put(MigrationStatus::Started(key));
+					read_write_ops.add_one_write();
+				}
+			}
+
+			read_write_ops
+		}
+
+		/// Tries to get the next key in the storage, returns None if there are no more keys to migrate.
+		/// Returns an error if the key is too long.
+		fn get_next_key(key: &StorageKey) -> (u64, NextKeyResult) {
+			if let Some(next) = sp_io::storage::next_key(key) {
+				let next: Result<StorageKey, _> = next.try_into();
+				match next {
+					Ok(key) => {
```

Comment:
```suggestion
					Ok(next_key) => {
```

reusing the same variable name as the parameter is dangerous, less readable and might introduce bugs